src/prefect/orion/database/migrations/versions/postgresql/2023_01_13_045500_2882cd2df463_implement_artifact_table.py

```diff
@@ -0,0 +1,207 @@
+        ),
+        sa.ForeignKeyConstraint(
+            ["flow_run_state_id"],
+            ["flow_run_state.id"],
+            name=op.f("fk_artifact__flow_run_state_id__flow_run_state"),
+        ),
+        sa.ForeignKeyConstraint(
+            ["task_run_id"],
+            ["task_run.id"],
+            name=op.f("fk_artifact__task_run_id__task_run"),
+        ),
+        sa.ForeignKeyConstraint(
+            ["task_run_state_id"],
+            ["task_run_state.id"],
+            name=op.f("fk_artifact__task_run_state_id__task_run_state"),
+        ),
+        sa.PrimaryKeyConstraint("id", name=op.f("pk_artifact")),
+    )
+    op.create_index(
+        op.f("ix_artifact__flow_run_id"), "artifact", ["flow_run_id"], unique=False
+    )
+    op.create_index(
+        op.f("ix_artifact__flow_run_state_id"),
+        "artifact",
+        ["flow_run_state_id"],
+        unique=False,
+    )
+    op.create_index(op.f("ix_artifact__key"), "artifact", ["key"], unique=True)
+    op.create_index(
+        op.f("ix_artifact__task_run_id"), "artifact", ["task_run_id"], unique=False
+    )
+    op.create_index(
+        op.f("ix_artifact__task_run_state_id"),
+        "artifact",
+        ["task_run_state_id"],
+        unique=False,
+    )
+    op.create_index(op.f("ix_artifact__updated"), "artifact", ["updated"], unique=False)
+
+    op.add_column(
+        "flow_run_state",
+        sa.Column(
+            "result_artifact_id", prefect.orion.utilities.database.UUID(), nullable=True
+        ),
+    )
+    op.create_index(
+        op.f("ix_flow_run_state__result_artifact_id"),
+        "flow_run_state",
+        ["result_artifact_id"],
+        unique=False,
+    )
+    op.create_foreign_key(
+        op.f("fk_flow_run_state__result_artifact_id__artifact"),
+        "flow_run_state",
+        "artifact",
+        ["result_artifact_id"],
+        ["id"],
+        ondelete="SET NULL",
+        use_alter=True,
+    )
+    op.add_column(
+        "task_run_state",
+        sa.Column(
+            "result_artifact_id", prefect.orion.utilities.database.UUID(), nullable=True
+        ),
+    )
+    op.create_index(
+        op.f("ix_task_run_state__result_artifact_id"),
+        "task_run_state",
+        ["result_artifact_id"],
+        unique=False,
+    )
+    op.create_foreign_key(
+        op.f("fk_task_run_state__result_artifact_id__artifact"),
+        "task_run_state",
+        "artifact",
+        ["result_artifact_id"],
+        ["id"],
+        ondelete="SET NULL",
+        use_alter=True,
+    )
+
+    with op.batch_alter_table("flow_run_state", schema=None) as batch_op:
+        batch_op.alter_column("data", new_column_name="_data")
+        batch_op.add_column(sa.Column("has_data", sa.Boolean))
+
+    with op.batch_alter_table("task_run_state", schema=None) as batch_op:
+        batch_op.alter_column("data", new_column_name="_data")
+        batch_op.add_column(sa.Column("has_data", sa.Boolean))
+
+    op.execute(
```

Comment:
1. This is a large update, we should do it in batches.
2. Is Postgres smart enough to use this index?

Reply:
I think this should be moved to its own migration and batched, probably

Reply:
The index is not necessary for backwards compatibility but required before the data migration